nemo/collections/nlp/parts/nlp_overrides.py

```diff
@@ -308,14 +333,145 @@ def update(self, new_scale=None):
+                        _scale,
+                        _growth_tracker,
+                        found_inf_combined,
+                        self._growth_factor,
+                        self._backoff_factor,
+                        self._growth_interval,
+                    )
+                else:
+                    # Only reset the growth tracker when hysteresis is larger than zero
+                    _growth_tracker.fill_(0.)
+            else:
+                # When no inf found, follow the native grad scale update rule.
+                # Increment growth_tracker, update scale when growth tracker reaches the interval, and
+                # reset the hysteresis tracker.
+                torch._amp_update_scale_(
+                    _scale,
+                    _growth_tracker,
+                    found_inf_combined,
+                    self._growth_factor,
+                    self._backoff_factor,
+                    self._growth_interval,
+                )
+                self._hysteresis_tracker = self.hysteresis
 
         # To prepare for next iteration, clear the data collected from optimizers this iteration.
         self._per_optimizer_states = defaultdict(torch.cuda.amp.grad_scaler._refresh_per_optimizer_state)
+
+    def state_dict(self):
+        """
+        Add hysteresis_tracker to the native functions' state_dict
+        """
+        return {"scale": self.get_scale(),
+                "growth_factor": self._growth_factor,
+                "backoff_factor": self._backoff_factor,
+                "growth_interval": self._growth_interval,
+                "_growth_tracker": self._get_growth_tracker(),
+                "_hysteresis_tracker": self._hysteresis_tracker} if self._enabled else {}
+
+    def load_state_dict(self, state_dict):
+        """
+        Load hysteresis_tracker in addition to the state dict of the native function
+        """
+        if not self._enabled:
+            return
+
+        if len(state_dict) == 0:
+            raise RuntimeError("The source state dict is empty, possibly because it was saved "
+                               "from a disabled instance of GradScaler.")
+
+        self._init_scale = state_dict["scale"]
+        if self._scale is not None:
+            self._scale.fill_(state_dict["scale"])
+        self._growth_factor = state_dict["growth_factor"]
+        self._backoff_factor = state_dict["backoff_factor"]
+        self._growth_interval = state_dict["growth_interval"]
+        self._init_growth_tracker = state_dict["_growth_tracker"]
+        if self._growth_tracker is not None:
+            self._growth_tracker.fill_(state_dict["_growth_tracker"])
+        self._hysteresis_tracker = state_dict["_hysteresis_tracker"]
+
+
+class MegatronHalfPrecisionPlugin(NativeMixedPrecisionPlugin):
+    """
+    Plugin for Half (FP16 and BF16) precision training.
+    This plugin assumes the uase of the optimizer with master parameters (fp32).
```

Comment:
typo: uase -> use

Reply:
Corrected the typo seems like they were there in Megatron repo.